File: apps/browser-extension/src/entrypoints/popup/utils/ConversionUtility.ts
import DOMPurify from 'dompurify';

/**
 * DOMPurify configuration for email viewing.
 * Allows safe HTML elements for email display while blocking XSS vectors.
 */
const EMAIL_SANITIZER_CONFIG = {
  ALLOWED_TAGS: [
    'div', 'span', 'p', 'br', 'hr',
    'h1', 'h2', 'h3', 'h4', 'h5', 'h6',
    'ul', 'ol', 'li',
    'table', 'thead', 'tbody', 'tfoot', 'tr', 'th', 'td',
    'a', 'img',
    'b', 'i', 'u', 's', 'strike', 'strong', 'em', 'small', 'sub', 'sup',
    'blockquote', 'pre', 'code',
    'font', 'center'
  ],
  ALLOWED_ATTR: [
    'style', 'class', 'id',
    'width', 'height', 'align', 'valign',
    'bgcolor', 'color', 'border',
    'cellpadding', 'cellspacing', 'colspan', 'rowspan',
    'face', 'size',
    'href', 'target', 'rel',
    'src', 'alt', 'title'
  ],
  ALLOW_DATA_ATTR: false,
  ADD_ATTR: ['target'],
  FORBID_TAGS: ['script', 'object', 'embed', 'iframe', 'frame', 'frameset',
    'form', 'input', 'button', 'textarea', 'select', 'option',
    'link', 'meta', 'base', 'applet'],
  FORBID_ATTR: ['onerror', 'onload', 'onclick', 'onmouseover', 'onmouseout',
    'onfocus', 'onblur', 'onchange', 'onsubmit', 'onreset', 'onkeydown',
    'onkeyup', 'onkeypress', 'ondblclick', 'oncontextmenu', 'onmousedown',
    'onmouseup', 'onmousemove', 'ondrag', 'ondrop']
};

/**
 * Utility class for conversion operations.
 * TODO: make this a shared utility class in root /core/ folder so we can reuse it between browser extension/mobile app
 * and possibly WASM client.
 */
class ConversionUtility {

  /**
   * Sanitizes HTML content for safe display in email viewers.
   * Removes all script tags, event handlers, and other XSS attack vectors.
   * @param html The HTML content to sanitize.
   * @returns Sanitized HTML safe for display.
   */
  public sanitizeHtmlForEmailViewing(html: string): string {
    if (!html || html.trim() === '') {
      return html;
    }

    try {
      return DOMPurify.sanitize(html, EMAIL_SANITIZER_CONFIG);
    } catch (ex) {
      console.error(`Error in sanitizeHtmlForEmailViewing: ${ex instanceof Error ? ex.message : String(ex)}`);
      // Return empty string on error to prevent potential XSS
      return '';
    }
  }

  /**
   * Sanitizes HTML content and converts anchor tags to open in a new tab.
   * This is a convenience method that combines sanitization with anchor tag conversion.
   * @param html The HTML content to process.
   * @returns Sanitized HTML with anchor tags configured to open in new tabs.
   */
  public sanitizeAndPrepareEmailHtml(html: string): string {
    if (!html || html.trim() === '') {
      return html;
    }

    // First sanitize to remove XSS vectors
    const sanitizedHtml = this.sanitizeHtmlForEmailViewing(html);

    // Then convert anchor tags to open in new tab
    return this.convertAnchorTagsToOpenInNewTab(sanitizedHtml);
  }

  /**
   * Convert all anchor tags to open in a new tab.
   * @param html HTML input.
   * @returns HTML with all anchor tags converted to open in a new tab when clicked on.
   *
   * Note: same implementation exists in c-sharp version in AliasVault.Shared.Utilities.ConversionUtility.cs
   */
  public convertAnchorTagsToOpenInNewTab(html: string): string {
    try {
      // Create a DOM parser
      const parser = new DOMParser();
      const doc = parser.parseFromString(html, 'text/html');

      // Select all anchor tags with href attribute
      const anchors = doc.querySelectorAll('a[href]');

      if (anchors.length > 0) {
        anchors.forEach((anchor: Element) => {
          // Handle target attribute
          if (!anchor.hasAttribute('target') || anchor.getAttribute('target') !== '_blank') {
            anchor.setAttribute('target', '_blank');
          }

          // Handle rel attribute for security
          if (!anchor.hasAttribute('rel')) {
            anchor.setAttribute('rel', 'noopener noreferrer');
          } else {
            const relValue = anchor.getAttribute('rel') ?? '';
            const relValues = new Set(relValue.split(' ').filter(val => val.trim() !== ''));

            relValues.add('noopener');
            relValues.add('noreferrer');

            anchor.setAttribute('rel', Array.from(relValues).join(' '));
          }
        });
      }

      return doc.documentElement.outerHTML;
    } catch (ex) {
      // Log the exception
      console.error(`Error in convertAnchorTagsToOpenInNewTab: ${ex instanceof Error ? ex.message : String(ex)}`);

      // Return the original HTML if an error occurs
      return html;
    }
  }

  /**
   * Normalize a username by converting it to lowercase and trimming whitespace.
   * @param username The username to normalize.
   * @returns The normalized username.
   */
  public normalizeUsername(username: string): string {
    return username.toLowerCase().trim();
  }
}

export default new ConversionUtility();

File: apps/server/Shared/AliasVault.Shared/Utilities/ConversionUtility.cs
//-----------------------------------------------------------------------
// <copyright file="ConversionUtility.cs" company="aliasvault">
// Copyright (c) aliasvault. All rights reserved.
// Licensed under the AGPLv3 license. See LICENSE.md file in the project root for full license information.
// </copyright>
//-----------------------------------------------------------------------

namespace AliasVault.Shared.Utilities;

using Ganss.Xss;
using HtmlAgilityPack;

/// <summary>
/// Class which contains various helper methods for data conversion.
/// </summary>
public static class ConversionUtility
{
    /// <summary>
    /// Lazy-initialized HTML sanitizer instance configured for safe email viewing.
    /// Removes all script tags, event handlers, and other XSS vectors while preserving
    /// safe HTML for email display.
    /// </summary>
    private static readonly Lazy<HtmlSanitizer> EmailSanitizer = new(() =>
    {
        var sanitizer = new HtmlSanitizer();

        // Allow common email formatting elements
        sanitizer.AllowedTags.Add("div");
        sanitizer.AllowedTags.Add("span");
        sanitizer.AllowedTags.Add("p");
        sanitizer.AllowedTags.Add("br");
        sanitizer.AllowedTags.Add("hr");
        sanitizer.AllowedTags.Add("h1");
        sanitizer.AllowedTags.Add("h2");
        sanitizer.AllowedTags.Add("h3");
        sanitizer.AllowedTags.Add("h4");
        sanitizer.AllowedTags.Add("h5");
        sanitizer.AllowedTags.Add("h6");
        sanitizer.AllowedTags.Add("ul");
        sanitizer.AllowedTags.Add("ol");
        sanitizer.AllowedTags.Add("li");
        sanitizer.AllowedTags.Add("table");
        sanitizer.AllowedTags.Add("thead");
        sanitizer.AllowedTags.Add("tbody");
        sanitizer.AllowedTags.Add("tfoot");
        sanitizer.AllowedTags.Add("tr");
        sanitizer.AllowedTags.Add("th");
        sanitizer.AllowedTags.Add("td");
        sanitizer.AllowedTags.Add("a");
        sanitizer.AllowedTags.Add("img");
        sanitizer.AllowedTags.Add("b");
        sanitizer.AllowedTags.Add("i");
        sanitizer.AllowedTags.Add("u");
        sanitizer.AllowedTags.Add("s");
        sanitizer.AllowedTags.Add("strike");
        sanitizer.AllowedTags.Add("strong");
        sanitizer.AllowedTags.Add("em");
        sanitizer.AllowedTags.Add("small");
        sanitizer.AllowedTags.Add("sub");
        sanitizer.AllowedTags.Add("sup");
        sanitizer.AllowedTags.Add("blockquote");
        sanitizer.AllowedTags.Add("pre");
        sanitizer.AllowedTags.Add("code");
        sanitizer.AllowedTags.Add("font");
        sanitizer.AllowedTags.Add("center");

        // Allow common styling attributes
        sanitizer.AllowedAttributes.Add("style");
        sanitizer.AllowedAttributes.Add("class");
        sanitizer.AllowedAttributes.Add("id");
        sanitizer.AllowedAttributes.Add("width");
        sanitizer.AllowedAttributes.Add("height");
        sanitizer.AllowedAttributes.Add("align");
        sanitizer.AllowedAttributes.Add("valign");
        sanitizer.AllowedAttributes.Add("bgcolor");
        sanitizer.AllowedAttributes.Add("color");
        sanitizer.AllowedAttributes.Add("border");
        sanitizer.AllowedAttributes.Add("cellpadding");
        sanitizer.AllowedAttributes.Add("cellspacing");
        sanitizer.AllowedAttributes.Add("colspan");
        sanitizer.AllowedAttributes.Add("rowspan");
        sanitizer.AllowedAttributes.Add("face");
        sanitizer.AllowedAttributes.Add("size");

        // Allow href for links but sanitize URLs
        sanitizer.AllowedAttributes.Add("href");
        sanitizer.AllowedAttributes.Add("target");
        sanitizer.AllowedAttributes.Add("rel");

        // Allow src for images but sanitize URLs
        sanitizer.AllowedAttributes.Add("src");
        sanitizer.AllowedAttributes.Add("alt");
        sanitizer.AllowedAttributes.Add("title");

        // Explicitly remove dangerous elements (fallback)
        sanitizer.AllowedTags.Remove("script");
        sanitizer.AllowedTags.Remove("object");
        sanitizer.AllowedTags.Remove("embed");
        sanitizer.AllowedTags.Remove("iframe");
        sanitizer.AllowedTags.Remove("frame");
        sanitizer.AllowedTags.Remove("frameset");
        sanitizer.AllowedTags.Remove("form");
        sanitizer.AllowedTags.Remove("input");
        sanitizer.AllowedTags.Remove("button");
        sanitizer.AllowedTags.Remove("textarea");
        sanitizer.AllowedTags.Remove("select");
        sanitizer.AllowedTags.Remove("option");
        sanitizer.AllowedTags.Remove("link");
        sanitizer.AllowedTags.Remove("meta");
        sanitizer.AllowedTags.Remove("base");
        sanitizer.AllowedTags.Remove("applet");

        return sanitizer;
    });

    /// <summary>
    /// Sanitizes HTML content for safe display in email viewers.
    /// Removes all script tags, event handlers, and other XSS attack vectors.
    /// </summary>
    /// <param name="html">The HTML content to sanitize.</param>
    /// <returns>Sanitized HTML safe for display.</returns>
    /// <remarks>
    /// This method should be called before displaying any untrusted HTML content
    /// (e.g., received emails) to prevent Cross-Site Scripting (XSS) attacks.
    /// </remarks>
    public static string SanitizeHtmlForEmailViewing(string html)
    {
        if (string.IsNullOrWhiteSpace(html))
        {
            return html;
        }

        try
        {
            return EmailSanitizer.Value.Sanitize(html);
        }
        catch (Exception ex)
        {
            // Log the exception
            Console.WriteLine($"Error in SanitizeHtmlForEmailViewing: {ex.Message}");

            // Return empty string on error to prevent potential XSS
            // This is safer than returning the original HTML
            return string.Empty;
        }
    }

    /// <summary>
    /// Sanitizes HTML content and converts anchor tags to open in a new tab.
    /// This is a convenience method that combines sanitization with anchor tag conversion.
    /// </summary>
    /// <param name="html">The HTML content to process.</param>
    /// <returns>Sanitized HTML with anchor tags configured to open in new tabs.</returns>
    public static string SanitizeAndPrepareEmailHtml(string html)
    {
        if (string.IsNullOrWhiteSpace(html))
        {
            return html;
        }

        // First sanitize to remove XSS vectors
        var sanitizedHtml = SanitizeHtmlForEmailViewing(html);

        // Then convert anchor tags to open in new tab
        return ConvertAnchorTagsToOpenInNewTab(sanitizedHtml);
    }

    /// <summary>
    /// Convert all anchor tags to open in a new tab.
    /// </summary>
    /// <param name="html">HTML input.</param>
    /// <returns>HTML with all anchor tags converted to open in a new tab when clicked on.</returns>
    /// <remarks>
    /// Note: same implementation exists in browser extension Typescript version in ConversionUtility.ts.
    /// </remarks>
    public static string ConvertAnchorTagsToOpenInNewTab(string html)
    {
        try
        {
            var doc = new HtmlDocument();
            doc.LoadHtml(html);

            var anchors = doc.DocumentNode.SelectNodes("//a[@href]");
            if (anchors != null)
            {
                foreach (var anchor in anchors)
                {
                    var targetAttr = anchor.Attributes["target"];
                    if (targetAttr == null)
                    {
                        anchor.SetAttributeValue("target", "_blank");
                    }
                    else if (targetAttr.Value != "_blank")
                    {
                        targetAttr.Value = "_blank";
                    }

                    // Add rel="noopener noreferrer" for security
                    var relAttr = anchor.Attributes["rel"];
                    if (relAttr == null)
                    {
                        anchor.SetAttributeValue("rel", "noopener noreferrer");
                    }
                    else if (!relAttr.Value.Contains("noopener") || !relAttr.Value.Contains("noreferrer"))
                    {
                        var relValues = new HashSet<string>(relAttr.Value.Split(' ', StringSplitOptions.RemoveEmptyEntries));
                        relValues.Add("noopener");
                        relValues.Add("noreferrer");
                        anchor.SetAttributeValue("rel", string.Join(" ", relValues));
                    }
                }
            }

            return doc.DocumentNode.OuterHtml;
        }
        catch (Exception ex)
        {
            // Log the exception
            Console.WriteLine($"Error in ConvertAnchorTagsToOpenInNewTab: {ex.Message}");

            // Return the original HTML if an error occurs
            return html;
        }
    }
}
